pkg/network/proxy/hybridproxier_test.go

```diff
@@ -0,0 +1,291 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+
+	unidlingapi "github.com/openshift/api/unidling/v1alpha1"
+)
+
+func makeService(namespace, name string) *corev1.Service {
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+			UID:       ktypes.UID(namespace+"/"+name),
+		},
+	}
+
+	return svc
+}
+
+func createServiceAndWait(svc *corev1.Service, kubeClient kubernetes.Interface, serviceLister corev1listers.ServiceLister) error {
+	_, err := kubeClient.CoreV1().Services(svc.Namespace).Create(context.TODO(), svc, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return utilwait.Poll(10*time.Millisecond, time.Second, func() (bool, error) {
+		_, err := serviceLister.Services(svc.Namespace).Get(svc.Name)
+		if err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+func deleteServiceAndWait(svc *corev1.Service, kubeClient kubernetes.Interface, serviceLister corev1listers.ServiceLister) error {
+	err := kubeClient.CoreV1().Services(svc.Namespace).Delete(context.TODO(), svc.Name, metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	return utilwait.Poll(10*time.Millisecond, time.Second, func() (bool, error) {
+		_, err := serviceLister.Services(svc.Namespace).Get(svc.Name)
+		if err == nil {
+			return false, nil
+		}
+		return true, nil
+	})
+}
+
+func TestHybridProxy(t *testing.T) {
+	mainProxy := newTestProxy("main", t)
+	unidlingProxy := newTestProxy("unidling", t)
+
+	kubeClient := fake.NewSimpleClientset()
+	kubeInformers := informers.NewSharedInformerFactory(kubeClient, time.Hour)
+
+	proxy := &HybridProxier{
+		mainProxy:     mainProxy,
+		unidlingProxy: unidlingProxy,
+		serviceLister: kubeInformers.Core().V1().Services().Lister(),
+
+		usingUserspace:      make(map[ktypes.NamespacedName]bool),
+		switchedToUserspace: make(map[ktypes.NamespacedName]bool),
+		pendingDeletion:     make(map[ktypes.NamespacedName]bool),
+	}
+
+	stopCh := make(chan struct{})
+	kubeInformers.Start(stopCh)
+	defer close(stopCh)
+
+	// *****
+
+	// Create a Service...
+	svc1 := makeService("testns", "one")
+	err := createServiceAndWait(svc1, kubeClient, proxy.serviceLister)
+	if err != nil {
+		t.Fatalf("unexpected error creating service: %v", err)
+	}
+	proxy.OnServiceAdd(svc1)
+
+	mainProxy.assertEvents("after creating first service",
+		"add service testns/one",
+	)
+	unidlingProxy.assertNoEvents("after creating first service")
+
+	// Then create its endpoints. Note that the unidling proxy is told about all
+	// endpoints changes, even though it doesn't know about the service
+	ep1 := makeEndpoints("testns", "one", "1.2.3.4")
+	proxy.OnEndpointsAdd(ep1)
+
+	mainProxy.assertEvents("after creating first endpoints",
+		"add endpoints testns/one 1.2.3.4",
+	)
+	unidlingProxy.assertEvents("after creating first endpoints",
+		"add endpoints testns/one 1.2.3.4",
+	)
+
+	// Idle the service; both the service and the endpoints will be removed from the
```

Comment:
didn't we move to label services instead of endpoints to support endpointslices too?

Reply:
Did we? We didn't _stop_ labeling Endpoints though, and we still have to track those anyway...
It would make the code cleaner if we could track idleness via Services, but that can happen later.